url7f15a160d2_timofei7_old_ai_assignments_tgz-pJ8-org_json_TestJ8/src/hw4/DomainList.java

```diff
@@ -0,0 +1,150 @@
+package hw4;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+/**
+ * Each variable has a domain.  Unary constraints are enforced by shrinking domains, and in a forward-checking
+ * or MAC algorithm, the domain is shrunk as a path in the backtrackign algorithm is explored.  We want to be able to add variables to the
+ * set in constant time, or (maybe more importantly, remove them), but we might also want to iterate over 
+ * list.  (Why?  We'll need to iterate to do a backchaining search, since the next value in the domain 
+ * needs to be selected).  So a LinkedHashSet is promising.  
+ * Notice that since Integers are immutable, we can freely clone domains
+ * @author tim
+ *
+ */
+public class DomainList implements Cloneable{
+
+	public ArrayList<LinkedHashSet<Integer>> domains;
+	public int numVariables;
+	
+	
+	/**
+	 * domainlist constructor
+	 * @param n number of variables
+	 * @param maxValue max number of values
+	 */
+	public DomainList(int n, int maxValue) {
+		numVariables = n; 
+		domains = new ArrayList<LinkedHashSet<Integer>>(n);
+		
+		for(int i = 0; i < n; i++ ) {
+			LinkedHashSet<Integer> dom = new LinkedHashSet<Integer>();
+			
+//			just make it empty fill in later
+//			for(int j = 0; j < maxValue; j++) { 
+//				dom.add(j);
+//			}
+			
+			domains.add(dom);
+		}
+	}
+
+	
+	/**
+	 * delete a value from a domain given a variable
+	 * @param variable
+	 * @param value
+	 */
+	public void deleteValue(int variable, int value)
+	{
+		LinkedHashSet<Integer> dom = domains.get(variable);
+		dom.remove(value);
+	}
+	
+	
+	/**
+	 * add a value into a domain given a variable
+	 * @param variable
+	 * @param value
+	 */
+	public void addValue(int variable, int value)
+	{
+		LinkedHashSet<Integer> dom = domains.get(variable);
+		dom.add(value);
+	}
+	
+	
+	/**
+	 * get the domain values for a variable
+	 * @param variable
+	 * @param value
+	 */
+	@SuppressWarnings("unchecked")
+	public Set<Integer> getValues(int variable)
+	{
+		return (LinkedHashSet<Integer>)(domains.get(variable).clone());
+	}
+	
+	/**
+	 * print out the domainlist
+	 */
+	public String toString() {
+		String s = "";
+		for(int i = 0; i < numVariables; i++) {
+			s = s + domains.get(i).toString() + "\n";
+		}
+		
+		return s;
+	}
+	
+	
+	
+	/**
+	 * uses the provided method to convert properly to a string
+	 * an exercise in java reflection mainly 
+	 * @param f the component object
+	 * @param c the converter method
+	 * @param aug a supplementary field for the converter if necessary
+	 * @return
+	 */
+	public String toString(Object f, Method c, int aug) {
+		String s = "";
+		for(int i = 0; i < numVariables; i++) {
+			try
+			{
+				for (Integer g : domains.get(i))
+				{
+					s = s + c.invoke(f, new Object[] { g, aug });
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)